From ddb4bca9dd3a7738d6087f4d4077ce489d59e7a5 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Tue, 20 Dec 2016 14:20:28 +0100
Subject: [PATCH 0668/1345] telephony: mvebu: rework tdm_if_exit

commit  6da98a07c8622e06e7ab963b54b646c0c9c99479 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

* Remove ifdefs and cleanup code
* Extract tdm2c stopping sequence to helper routine tdm2c_pcm_disable()

Change-Id: Ic3d1c136c47884db4543b01c6ef2eb34df89c0bf
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35037
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone_dev.c |   66 ++++++++++++++++----------
 1 file changed, 40 insertions(+), 26 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index fcf8154..8c56d213 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -434,6 +434,23 @@ int tdm_if_init(struct tal_params *tal_params)
 	return ret;
 }
 
+void tdm2c_pcm_disable(void)
+{
+	u32 max_poll = 0;
+
+	tdm2c_if_pcm_stop();
+
+	while ((is_pcm_stopping != 0) && (max_poll < 20)) {
+		mdelay(1);
+		max_poll++;
+	}
+
+	if (max_poll >= 20)
+		dev_warn(priv->dev, "%s: stopping pcm channels exceeded 20ms\n",
+			 __func__);
+
+}
+
 void tdm_if_exit(void)
 {
 	int i;
@@ -443,37 +460,34 @@ void tdm_if_exit(void)
 		return;
 
 	/* Stop PCM channels */
-	if (pcm_enable)
-#ifdef CONFIG_MV_TDM2C_SUPPORT
-		tdm2c_if_pcm_stop();
-#endif
-#ifdef CONFIG_MV_TDMMC_SUPPORT
-		tdmmc_if_pcm_stop();
-#endif
-
-#ifdef CONFIG_MV_TDM2C_SUPPORT
-		if (tdm_if_unit_type_get() == MV_TDM_UNIT_TDM2C) {
-			u32 max_poll = 0;
-
-			while ((is_pcm_stopping != 0) && (max_poll < 20)) {
-				mdelay(1);
-				max_poll++;
-			}
-
-			if (max_poll >= 20)
-				dev_warn(priv->dev, "%s: waiting for pcm channels to stop exceeded 20ms\n", __func__);
+	if (pcm_enable) {
+		switch (priv->tdm_type) {
+		case MV_TDM_UNIT_TDM2C:
+			tdm2c_pcm_disable();
+			break;
+		case MV_TDM_UNIT_TDMMC:
+			tdmmc_if_pcm_stop();
+			break;
+		default:
+			dev_err(&priv->parent->dev, "%s: undefined TDM type\n",
+				__func__);
 		}
-#endif
+	}
 
+	/* Disable TDM and release resources */
 	if (tdm_init) {
-#ifdef CONFIG_MV_TDM2C_SUPPORT
-		if (tdm_if_unit_type_get() == MV_TDM_UNIT_TDM2C)
+		switch (priv->tdm_type) {
+		case MV_TDM_UNIT_TDM2C:
 			tdm2c_release();
-#endif
-#ifdef CONFIG_MV_TDMMC_SUPPORT
-		if (tdm_if_unit_type_get() == MV_TDM_UNIT_TDMMC)
+			break;
+		case MV_TDM_UNIT_TDMMC:
 			tdmmc_release();
-#endif
+			break;
+		default:
+			dev_err(&priv->parent->dev, "%s: undefined TDM type\n",
+				__func__);
+		}
+
 		/* Remove proc directory & entries */
 		remove_proc_entry("tdm_stats", tdm_stats);
 		remove_proc_entry("tdm", NULL);
-- 
1.7.9.5

